cmd/config/docs/api-conventions/functions-spec.md

```diff
@@ -18,168 +23,359 @@ interpreted as described in [RFC 2119][2].
 - Low-level tweaks to the output of high-level abstractions
-  - e.g. add an `init container` to a client _"CRD"_ Resource after it was generated.
+  - e.g. add an `init container` to a client _"CRD"_ resource after it was
+    generated.
 - Composition and layering of multiple functions together
   - Compose generation, injection, validation together
 
-## Spec
+## Definitions
 
-### Input Type
+- **function:** A containerized program conforming to the spec described in this
+  document.
+- **orchestrator:** A program that invokes the function container, passing
+  arguments and processing its output.
 
-A function MUST accept as input a single [Kubernetes List type][3].
-The `items` field in the input will contain a sequence of [Object types][3].
-A function MAY not support [Simple types][3] and List types.
+## Interface
 
-An example using `v1/ConfigMapList` as input:
+The inter-process communication between the orchestrator and a function works as
+follows:
 
-```yaml
-apiVersion: v1
-kind: ConfigMapList
-items:
-  - apiVersion: v1
-    kind: ConfigMap
-    metadata:
-      name: config1
-    data:
-      p1: v1
-      p2: v2
-  - apiVersion: v1
-    kind: ConfigMap
-    metadata:
-      name: config2
-```
+1. Orchestrator runs the function container and provides the input on `stdin`.
+   The input is a Kubernetes object of kind `ResourceList` as described below.
+2. Function reads the input from `stdin`, performs computations, and provides
+   the output as a `ResourceList` to `stdout`. The function MAY also emit
+   non-structured error message on `stderr`.
+3. Orchestrator uses the `stdout`, `stderr`, and the exit code of the function
+   as it sees fit following to the semantics described below.
+
+### Schema
 
-An example using `v1/List` as input:
+A function MUST accept input from `stdin` and MUST output to `stdout` a
+Kubernetes object of kind `ResourceList` with the following OpenAPI schema:
 
 ```yaml
-apiVersion: v1
-kind: List
-items:
-  - apiVersion: foo-corp.com/v1
-    kind: FulfillmentCenter
-    metadata:
-      name: staging
-    address: "100 Main St."
-  - apiVersion: rbac.authorization.k8s.io/v1
-    kind: ClusterRole
-    metadata:
-      name: namespace-reader
-    rules:
-      - resources:
-          - namespaces
-        apiGroups:
-          - ""
-        verbs:
-          - get
-          - watch
-          - list
+swagger: "2.0"
+info:
+  title: KRM Functions Specification (ResourceList)
+  version: v1
+definitions:
+  ResourceList:
+    type: object
+    description: ResourceList is the input/output wire format for KRM functions.
+    x-kubernetes-group-version-kind:
+      - group: config.kubernetes.io
+        kind: ResourceList
+        version: v1
+      - group: config.kubernetes.io
+        kind: ResourceList
+        version: v1beta1
+    required:
+      - items
+    properties:
+      apiVersion:
+        description: apiVersion of ResourceList
+        type: string
```

Comment:
Is there a way to say that only allowed value here is: `config.kubernetes.io/v1` ?

I remember in the openAPI schema section of k8s CRDs, one could specify these validation properties in a declarative way.

I was thinking of leveraging this schema for validating the `input/output` from a function in the orchestrator. Same logic can be even packaged as an independent function `fn-spec-validator` that can be chained to validate if the output from my function is compliant or not and can be integrated in the SDKs.